systemd: add systemd-importd

wucke13 opened the pull request:
###### Motivation for this change

Adding `systemd-importd` to the build, so that `machinectl`s `import-.*`
may actually do anything.  Currently they fail with

```
Failed to transfer image: The name org.freedesktop.import1 was not provided by any .service files
```
as `systemd-importd` is not built.
This potentially will fix #70348

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->

###### Testing

1. spawn a nixos vm based on this branch with `gnupg` and `curl` in the system environment and working `sudo`
2. run this (as normal user with working `sudo`) :
```
sudo gpg --no-default-keyring --keyring=/etc/systemd/import-pubring.gpg --fingerprint
tfile=$(mktemp -u /tmp/masterkey.nspawn.org.XXXXXXXXXXX)
curl https://nspawn.org/storage/masterkey.pgp -o "$tfile"
sudo gpg --no-default-keyring --keyring=/etc/systemd/import-pubring.gpg --import "$tfile"
machinectl pull-tar https://nspawn.org/storage/ubuntu/bionic/tar/image.tar.xz
machinectl start image
machinectl shell image
```
You will get a lot of selinux warning (which is perfectly fine if your vm has no se linux). In the end, the image should be verified by checksum and the signature proven by the imported key.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

wucke13 committed:
systemd: adding support for systemd-importd

Adding `systemd-importd` to the build, so that `machinectl`s `import-.*`
may actually do anything. Currently they fail with

```
Failed to transfer image: The name org.freedesktop.import1 was not provided by any .service files
```
as `systemd-importd` is not built. Also registers the regarding dbus
api and service in the systemd module.

andir commented:
I'll try to have a look at this in the next days. I vaguely remember reasons for not doing what you did there with udev. I can't remember right now. Hopefully that gets back to me within the next days.

wucke13 commented:
pin @andir

edolstra commented:
What's the effect on closure size? It doesn't seem desirable to me to make systemd dependent on X.org (via libxkbdcommon).

wucke13 commented:
> What's the effect on closure size?

How would I find out?

Regarding `libxkbdcommon`: I have no clue why I added it, but it seems to be not necessary. I removed the unneeded deps and rebased on master. I will test 7d20b67 it in a vm and report back in a few hours.

edolstra commented:
> How would I find out?

`nix path-info -S` (or `-rS` to see the paths in the closure).

wucke13 commented:
Ah, with the path to the nix store. I always tried `nix path-info -S systemd`. The closure size becomes `147.0M`, so round about `+35M`.
While testing, I found out that there is an issue with direct calls to `gpg` and `tar` in the import code, though that should be fixed in the next push.

edolstra commented:
Hm, I'd really prefer to avoid a hard dependency on X11 or gpg in the base system.

wucke13 commented:
That seems reasonable. Do we have any other `systemd-*` component which is seperated from the `systemd` base expression?

andir commented:
Is it feasible to have those additional binaries in another output without increasing the closure of the `out` output?

wucke13 commented:
> Is it feasible to have those additional binaries in another output without increasing the closure of the out output?

I'm afraid that it might become difficult marrying both parts together. I will look into it though.

>  another systemd package that carries all the extra bloat that most/many users do not care about.

You do realize that the current state is shipping not working bloat (at least `machinectl`) in the normal package? Of course having finer control over what to build and whatnot is good, but the user gets __bloat+regressions__ right now (of which at least the regressions shall go IMO). Maybe a general decision should be made on how to deal with `systemd`, IMHO the current package ships a lot of things which may be considered bloat since they are not at all needed for an `init`.

shibumi commented:
Please merge, I need this.

edolstra commented:
I don't think NixOS has an obligation to enable every systemd feature. Just because upstream feels that systemd needs to be docker replacement doesn't mean we *have* to impose that bloat on our users.

> What is it that makes being dependent on gpg bad?

Let's have a look at the dependencies of gpg:
```
$ nix path-info -rs nixpkgs#gnupg
/nix/store/0m9rsg8dqsr7pq0i42qr6in8lfrqqyvn-libcap-2.27-lib                   26936
/nix/store/0mccp18aa10j1v1lh3z3ajf1ivhm7c3j-libsecret-0.18.8                 594888
/nix/store/1a037dsgz0615j1hf9nshmpykc7qj39l-libtasn1-4.14                     89824
/nix/store/1k9nbcb6avap3c8jcnvabkdgj86wj3vv-libXdmcp-1.1.3                    32920
/nix/store/209bk6hiw0vdlg465qirycz59f6l0ls1-libICE-1.0.9                     118400
/nix/store/22mmqqgdhzrs7l9vv324knbsz4zbwycp-libksba-1.3.5                    262496
/nix/store/234i6czbva9dzwv3y4w0z9rr5p72m4g6-npth-1.6                          49792
/nix/store/2s3p317phvigx5zad2z7drvbspp385b9-libevent-2.1.11                  896840
/nix/store/3s5i0n6575jcbr149qnqwmqd12d4z737-libXfixes-5.0.3                   35416
/nix/store/4fl608fn4wfijpwp9dyiq3jm94v5s1yn-pixman-0.38.4                    799944
/nix/store/4kcacmw8klsv7fb5vb4fbp3wpaygmrbx-fribidi-1.0.5                    236256
/nix/store/4lyz1axsd4qqjzy580nxwpag476iagjd-pcsclite-1.8.25                   89760
/nix/store/51ac78a3jszrq4iss9fs5rlr1w1hg39c-pinentry-1.1.0                   323536
/nix/store/5573j8178kmyz3qxsvx25snrn0i92ghk-cyrus-sasl-2.1.27                622368
/nix/store/5dn03xfw3wimm78nfpg67p1zj1025sbc-bzip2-1.0.6.0.1                   80912
/nix/store/62ar9xmrlcnlgmwgfi77xz6bq1180vhi-zlib-1.2.11                      132608
/nix/store/63jwkphcgl7yzcx168yns0540f2c4msm-keyutils-1.6-lib                  42016
/nix/store/690qkhaj8mvr6ax949gz16a69wn5fcxs-libX11-1.6.8                    2780096
/nix/store/6friwcmy1yck5jh7l2nrd1503igafqsg-libpng-apng-1.6.37               256424
/nix/store/6hfmcn72gh0nd9dws8zyjy6264v0la9n-libXft-2.3.3                     123504
/nix/store/6yaj6n8l925xxfbcd65gzqx3dz7idrnn-glibc-2.27                     28147008
/nix/store/7hmm5sawx4mbb9jkczjgrn0gm96681qy-libtool-2.4.6-lib                 58080
/nix/store/7yjvankj7vvi22i2maf4crvpri5563jn-pcre-8.43                        497624
/nix/store/87qbb6wv1mk8a79qlzkhx8ql85l4npnn-graphite2-1.3.6                  307960
/nix/store/8kaqqz7x3fgyy84wn2zpbpkmdzvrm6x6-p11-kit-0.23.16.1               3643704
/nix/store/9sbsiqlsp3lclcksz1p5qjki2maamail-libXinerama-1.1.4                 22920
/nix/store/a2gqcm6dkr44x3dh370pbhdg63hh38cb-gtk+-2.24.32                   28310240
/nix/store/agc9nn5r0jm399lsfm0rx3m1gv40frsr-libunistring-0.9.10             1634416
/nix/store/aldz750xb5k96mzxg4l29dsip28b376k-glib-2.60.7                    12582496
/nix/store/bkpfgh57ijkfigx8kagiv9svr3cch77y-libXi-1.7.10                      82344
/nix/store/cbcbgz17zdhbsbj8fvmqv6hf6y8p13sx-pango-1.43.0                    1537968
/nix/store/ddx5p19qh2wdh7f7h7g5xm4cgsq6dnra-libSM-1.2.3                       46752
/nix/store/djsipbpa5v8pnspcjdwfqw805y3r8qkf-atk-2.32.0                      1304816
/nix/store/fgjjmczbhr6m043p14yikshxah0iwnd9-gdk-pixbuf-2.38.1               3105344
/nix/store/fgmkgmbmir4s2581w3m11idij649szdi-avahi-0.7                       1751272
/nix/store/fiac1v6mm0vhqw5qdjl6fkdzy21adv84-dbus-1.12.16-lib                 451368
/nix/store/h8gj0h3kjf2p449ks9yx9mq3k8ml0zpv-cups-2.2.12-lib                 3838272
/nix/store/hh2n0fhhwkyc01285wry1krlmyljn0dg-fontconfig-2.12.6                 93008
/nix/store/hir1khmcll26mjw30lvabn46bjwi61m9-libXcomposite-0.4.5               26216
/nix/store/hl3arlx4r35d6ygg68zix49k4z66gdlc-dejavu-fonts-minimal-2.37        757824
/nix/store/i945gbylyzlz5gcdsj81vfbaz9yji37w-lz4-1.9.1                        496608
/nix/store/im6s5xh2rnl9mp7fd11g8b4lhhlxpc8h-openldap-2.4.48                 4203992
/nix/store/isg5d6rh4zdva16hxs7qg6k1s96v39mx-sqlite-3.28.0                   1245824
/nix/store/j2p1qjbajqfb95ba3dqgjbsknipknikk-gmp-6.1.2                        698784
/nix/store/jfwhd03syvpaclfphj41zyaf0lrhp2hz-cairo-1.16.0                    1919704
/nix/store/jssgml4p8vzs7ic5cqy7hh48m57swdb9-dns-root-data-2019-01-11           4480
/nix/store/jsx6f89kxy4apq8w9y434rf5mjlsawzi-nettle-3.4.1                     513144
/nix/store/kc2577vmldwr2546k9p7nj5h8r5ar318-pcsclite-1.8.25-bin              151224
/nix/store/kci1jrp2m4srzgypfyk4flh1s6r622zx-libusb-1.0.23                    136688
/nix/store/kglyaag9larw9hc1pc7iybzj1yzyy5x2-libassuan-2.5.3                  100472
/nix/store/kszi9vwy4ra7zya9gnqg3rgla81mkcx0-util-linux-2.33.2               1658096
/nix/store/kvia4rwy9y4wis4v2kb9y758gj071p5v-ncurses-6.1-20190112            3488416
/nix/store/l6h4ya0wzb4b8mr0y58k2gh2nhfql4sn-bash-4.4-p23                    1321096
/nix/store/l9g11dsdj97d5q0747bqjacrakyrf8dk-db-5.3.28                       4119360
/nix/store/n7i4v3zz3xqg94w2wavdxj77g3gr18if-expat-2.2.7                      297352
/nix/store/nair5j8i8p9ik286ijkhyaip9nychsvy-libjpeg-turbo-2.0.3             1503872
/nix/store/najvrr9mppdrgjh529l79w8px36wn58z-gcc-8.3.0-lib                   5719760
/nix/store/nrnnvdy72487sv5n83qa04lafbl6nbcl-openssl-1.1.1c                  4188016
/nix/store/nw2xbfq1lb797fhv9w44in6cvwwyncr5-libglvnd-1.0.0                  2157760
/nix/store/pdpgm7h0mcbc9xidn0x1xyzcbqdzn2ad-libgcrypt-1.8.5                 1261064
/nix/store/pgnn8s912q6ps1byyp7kg3prw8hz8yv1-libselinux-2.9                   592312
/nix/store/q1k669np0apc4h48aikwa771gzl8d02n-libXrender-0.9.10                 55344
/nix/store/q8796x08wb1cmzqg5b6mx6k1k0kcxp8i-libgpg-error-1.36                718080
/nix/store/qq4jlc0q1spcgahjq2rinwqfx7hiwyc8-gnupg-2.2.17                   11268976
/nix/store/r0hyzrwj90mk0cxr5lrxxkcnl28j1g8d-unbound-1.9.3-lib                902928
/nix/store/r8q7p0jm2b1wg6rncr7y0fz0v39xhjam-gnutls-3.6.9                    2838224
/nix/store/ramhsw6khg128qvysraqlagcw8klg1ii-libXau-1.0.9                      24584
/nix/store/sc3wwzzy17xnzm2d9m5ydv098k6c2vy6-systemd-243-lib                 1710104
/nix/store/v0dy9gcr12vf63lyq69nrpgzm3yz5wlz-freetype-2.10.1                 1951008
/nix/store/w0zf7fmfs5ml7csscrmd0m89s0wzmr5x-libkrb5-1.17                    2389336
/nix/store/w1bkhjdgxx1q8qf0wfkdhkz50ygm0rra-libXext-1.3.4                     96960
/nix/store/wmlvk1gim9plg5719i02zykyajlka72y-libtiff-4.0.10                   568144
/nix/store/xm3qkbj31z8jza3a777wa7wxcg6c578l-libXcursor-1.2.0                  66688
/nix/store/y1kwd5gvczaw7if610gvcqci5fiwn5zg-harfbuzz-2.6.1                  1624096
/nix/store/yk3aqhhw3jj0sj9nzqs7vhjx9cwjlb78-fontconfig-2.12.6-lib            328248
/nix/store/yk8lp0x7sml42gaf9gfkr6cwx99glz7b-xz-5.2.4                         341056
/nix/store/ymb2qb76a6mr2gnn848yyfyrmn35kj6a-jasper-2.0.16                    985488
/nix/store/ynmrvr48grbwcr9nhv80indfndzwyg4r-libxcb-1.13.1                   1400072
/nix/store/ypdfsafs91j486cvlqi7p02a3rglakb0-libdaemon-0.14                    70760
/nix/store/z39757pg1gp5lgkxcn0yv6nv4lgmpnad-libffi-3.2.1                      51368
/nix/store/zd97p80q3nb662128ildvscnzmk0mrvc-libXrandr-1.5.2                   64928
/nix/store/zpav4lfjikaz21qkyq3hnaxsa9a7kqzy-readline-6.3p08                  396016
```
So depending on gpg means pulling in X11 libraries, Cairo, fonts, Avahi, GTK+, OpenLDAP and a bunch of things where I have no idea what they do. (This mostly seems to be because of gpg-agent / pinentry, a program that doesn't work for me half of the time...)

Of course this is all transitive, so any app in Nixpkgs that has a (transitive) dependency on systemd will now depend on those things. (Though usually it's a dependency on `systemd.lib` so maybe it's not too bad.)

Lassulus commented:
isn't the gnupg build here build without guiSupport, so the x11 dependency would go away?
also, can't we hide this feature behind an enable flag in the systemd derivation and people could enable it via an option?

wucke13 commented:
> Just because upstream feels that systemd needs to be docker replacement doesn't mean we have to impose that bloat on our users.

__I can not stress this enough: nixos already ships `machinectl`, `nspawn` & `machined` by default. You already ship this bloat, but with a regression, as you do no ship the single component which is missing to make `machined`/`machinectl` any usable, because there is no official way of importing images! If you wish to make an argument about this then stop shipping broken bloat!__

> Let's have a look at the dependencies of gpg:

Let's have a look at the dependencies of `systemd` from this pr:

```
/nix/store/08al8hwj0hvqahb0yibslsmsvj50b1qz-acl-2.2.53                  	     109920
/nix/store/0m9rsg8dqsr7pq0i42qr6in8lfrqqyvn-libcap-2.27-lib             	      26936
/nix/store/0ni6d49nci8vpzxr7672qjwr6f5c5kc4-util-linux-2.33.2           	    1658096
/nix/store/0waqa7f6l0lzxk815ir633gaq0mas445-libidn2-2.2.0               	     214360
/nix/store/1a037dsgz0615j1hf9nshmpykc7qj39l-libtasn1-4.14               	      89824
/nix/store/22mmqqgdhzrs7l9vv324knbsz4zbwycp-libksba-1.3.5               	     262496
/nix/store/234i6czbva9dzwv3y4w0z9rr5p72m4g6-npth-1.6                    	      49792
/nix/store/2p33f8vmb094grg691f8w50gzpp7x884-libnftnl-1.1.4              	     280128
/nix/store/38jvvx3z7cz8k58v535r4sbswgp07yaw-cracklib-2.9.7              	    9971448
/nix/store/440vkrx6mf7v4j4z88vjcp6pavk05lig-bzip2-1.0.6.0.1-bin         	      70016
/nix/store/4ad65dga6w342wh9brlxdxwca0l31n33-audit-2.8.5                 	     311504
/nix/store/5dn03xfw3wimm78nfpg67p1zj1025sbc-bzip2-1.0.6.0.1             	      80912
/nix/store/62ar9xmrlcnlgmwgfi77xz6bq1180vhi-zlib-1.2.11                 	     132608
/nix/store/63jwkphcgl7yzcx168yns0540f2c4msm-keyutils-1.6-lib            	      42016
/nix/store/6qggmn4di5rhylh9jlah6rbcz6vwhb2x-libevent-2.1.11             	     896840
/nix/store/6yaj6n8l925xxfbcd65gzqx3dz7idrnn-glibc-2.27                  	   28147008
/nix/store/7hmm5sawx4mbb9jkczjgrn0gm96681qy-libtool-2.4.6-lib           	      58080
/nix/store/8kaqqz7x3fgyy84wn2zpbpkmdzvrm6x6-p11-kit-0.23.16.1           	    3643704
/nix/store/8wr3wcj9qldb2ak8ccxvyf4dm9zzh7cx-libseccomp-2.4.1-lib        	     305720
/nix/store/9pfanxsm4wpygrx13r9332lsmhbqnmmj-gnupg-2.2.17                	   11221672
/nix/store/a66s8hzy6w34dlh6ahgfdwaiglpcny65-gnutar-1.32                 	    2858776
/nix/store/agc9nn5r0jm399lsfm0rx3m1gv40frsr-libunistring-0.9.10         	    1634416
/nix/store/b3pr9iv2v3j1y0gvdhy86pq5s9lqgzg7-openldap-2.4.48             	    4203992
/nix/store/b6fp8g085zyhq02vivdzdirb4mgp5ggi-openssl-1.1.1d              	    4158648
/nix/store/bswbcw4y9mr4ngj6r7gsv3drqlcnpldf-libkrb5-1.17                	    2389336
/nix/store/cc6ilhxss244dy7lqs6573bn8xp4wpq9-systemd-243-lib             	    1710104
/nix/store/d9mj4lwxcibgdm5aaa66r8i9svnb6bpm-linux-pam-1.3.1             	    2004888
/nix/store/g0d5ykr6jxbl4h7rhsz3mi5hp7gz4p2y-kbd-2.0.4                   	    2618360
/nix/store/g6hcw252yn4shagc7wg40biy3hmmpn09-libmnl-1.0.4                	      43704
/nix/store/gasyimps4d9lwl1pc44mhzgjb20dm7i6-libnetfilter_conntrack-1.0.7	     205256
/nix/store/hljwhpqjhpy5327zfmzjjh8xyly3hiy2-glibc-2.27-bin              	    2704872
/nix/store/i945gbylyzlz5gcdsj81vfbaz9yji37w-lz4-1.9.1                   	     496608
/nix/store/isg5d6rh4zdva16hxs7qg6k1s96v39mx-sqlite-3.28.0               	    1245824
/nix/store/j2p1qjbajqfb95ba3dqgjbsknipknikk-gmp-6.1.2                   	     698784
/nix/store/ji9yrngymyigsqzcqaavng7v4dws32q7-nghttp2-1.39.2-lib          	     193176
/nix/store/jssgml4p8vzs7ic5cqy7hh48m57swdb9-dns-root-data-2019-01-11    	       4480
/nix/store/jsx6f89kxy4apq8w9y434rf5mjlsawzi-nettle-3.4.1                	     513144
/nix/store/kglyaag9larw9hc1pc7iybzj1yzyy5x2-libassuan-2.5.3             	     100472
/nix/store/kn9w6k46qmzrv3hgzs1csrrs3dcad7h6-attr-2.4.48                 	      89136
/nix/store/kvia4rwy9y4wis4v2kb9y758gj071p5v-ncurses-6.1-20190112        	    3488416
/nix/store/kx56d14p0qm2awn0qrqrv4rb8xdhvhpi-coreutils-8.31              	    1798432
/nix/store/l6h4ya0wzb4b8mr0y58k2gh2nhfql4sn-bash-4.4-p23                	    1321096
/nix/store/l9g11dsdj97d5q0747bqjacrakyrf8dk-db-5.3.28                   	    4119360
/nix/store/lng218nwk66q5l3c3pq4826ggj4rvyfr-libmicrohttpd-0.9.66        	     167744
/nix/store/m52arbrnwm1km3yagr9iamsp2sqbmmip-libssh2-1.9.0               	     290464
/nix/store/mig9a44n4swlhwyamy8hhghnh4ifxk4j-unbound-1.9.3-lib           	     902928
/nix/store/mvs0vqiqp6186wpgs7z18agw0gs6l6ji-pcre2-10.33                 	    1672000
/nix/store/najvrr9mppdrgjh529l79w8px36wn58z-gcc-8.3.0-lib               	    5719760
/nix/store/pdpgm7h0mcbc9xidn0x1xyzcbqdzn2ad-libgcrypt-1.8.5             	    1261064
/nix/store/pk6h6pzmwaclxb9rpzvh5rk346mq8nyl-iptables-1.8.3              	    2936360
/nix/store/q1b45q29p5vw5bvq7p8prq94sgxnqyx3-systemd-243                 	   24236288
/nix/store/q8796x08wb1cmzqg5b6mx6k1k0kcxp8i-libgpg-error-1.36           	     718080
/nix/store/q9vxib6xc66i3w9pqyddbdlhqc4rshc3-kexec-tools-2.0.20          	     243424
/nix/store/qira6x24xinnzg3ksgz1znj0k0i04v6i-gnutls-3.6.9                	    2838224
/nix/store/r4i08s94s97jiwx5pwnw2qzgdwi404vn-libapparmor-2.13.1          	     445640
/nix/store/r5s3w32ahjzdlzsfrhybc3l2qcpi6yb2-libpcap-1.9.0               	    1028920
/nix/store/sgganmpm2bbnqmgz35hrgp4d0cvml6ir-gzip-1.10                   	     150120
/nix/store/wx84sq1p3ji0va77hchhfr1d6jd1b24h-cyrus-sasl-2.1.27           	     622368
/nix/store/x0yc4mysvi3p0k46wiipd4v7n6g4kwb4-libnfnetlink-1.0.1          	      57400
/nix/store/xnhkyg25aw87ivf7adbnfivfg779a5aw-util-linux-2.33.2-bin       	    5676928
/nix/store/yk8lp0x7sml42gaf9gfkr6cwx99glz7b-xz-5.2.4                    	     341056
/nix/store/ynyacrpj8876kqv7xmbhak1w2gmp4rk4-curl-7.65.3                 	     641560
/nix/store/z0bpm56favz7yzas3j9023lqisbr4q5s-kmod-26                     	     329368
/nix/store/z39757pg1gp5lgkxcn0yv6nv4lgmpnad-libffi-3.2.1                	      51368
/nix/store/z7y1hqcw0h6x5pm9piilxpv68v0s6n31-shadow-4.6                  	    3794936
/nix/store/zpav4lfjikaz21qkyq3hnaxsa9a7kqzy-readline-6.3p08             	     396016
/nix/store/zrd8dbxrq8shpfy2xh68l02qqyd4vy97-db-4.8.30                   	    3494896
```

Here is a diff between the systemd actually running in my systemd and the one in my branch:

```
diff systemd-normal systemd-importd
8a9,10
> curl-7.65.3                 	 626.5K
> cyrus-sasl-2.1.27           	 607.8K
9a12
> db-5.3.28                   	   3.9M
14a18,19
> gnupg-2.2.17                	  10.7M
> gnutar-1.32                 	   2.7M
19a25
> keyutils-1.6-lib            	  41.0K
21a28
> libassuan-2.5.3             	  98.1K
27a35,36
> libkrb5-1.17                	   2.3M
> libksba-1.3.5               	 256.3K
34a44
> libssh2-1.9.0               	 283.7K
35a46
> libtool-2.4.6-lib           	  56.7K
38a50
> ncurses-6.1-20190112        	   3.3M
39a52,55
> nghttp2-1.39.2-lib          	 188.6K
> npth-1.6                    	  48.6K
> openldap-2.4.48             	   4.0M
> openssl-1.1.1d              	   4.0M
41a58
> readline-6.3p08             	 386.7K
43c60,61
< systemd-243                 	  22.8M
---
> sqlite-3.28.0               	   1.2M
> systemd-243                 	  23.1M
45c63
< unbound-1.9.4-lib           	 881.8K
---
> unbound-1.9.3-lib           	 881.8K
```

As `gpg` is only used to verify signatures, I can look into making the `gpg` expression used even more minimal, maybe I can get rid of `openldap` too. Would that make you less opposed to this pr?

> isn't the gnupg build here build without guiSupport, so the x11 dependency would go away?

Yes!

> also, can't we hide this feature behind an enable flag in the systemd derivation and people could enable it via an option?

That might be something to do in the future, but I see two problems here: 
+ the current `systemd` ships a broken component. It ain't clever IMHO to ship something partially broken by default, and give an opt-in to fix it.
+ also it might cause some headache to have two `systemd` packages. If it is a switch in the expression, people who alter it will have two different versions of `systemd` at all time, since many packages depend on the `systemd` with the switch set to false. On the other hand, if the switch is global there will be two versions of all packages which depend on `systemd`.

edolstra commented:
>  I can not stress this enough: nixos already ships machinectl, nspawn & machined by default.

Because they are useful even without systemd-importd. E.g. `nixos-container` uses them. Also, the bloat doesn't come from those programs but from dependencies like `gnupg`.

> isn't the gnupg build here build without guiSupport, so the x11 dependency would go away?

Using `guiSupport = false` is actually likely to lead to an *increase* in system closure size, because many users will end up with 2 builds of `gnupg` in their closure.

It would be better if `pinentry` was architected like `systemd-ask-password`, which doesn't require programs to have a direct dependency on some GUI components to show a password prompt.

wucke13 commented:
> Just because upstream feels that systemd needs to be docker replacement doesn't mean we have to impose that bloat on our users.

gave me another impression than

> Because they are useful even without systemd-importd. E.g. nixos-container uses them. Also, the bloat doesn't come from those programs but from dependencies like gnupg.

Thank you for explaining that. 

> Using guiSupport = false is actually likely to lead to an increase in system closure size, because many users will end up with 2 builds of gnupg in their closure.

Only for those who use `gpg`, and then it's about +10Mbyte. But doing that decreases the deps of `systemd` by a lot (to the point that `systemd` from this pr has less deps in total than normal `gnupg` alone). I gott that you have both an issue with depending on any form of `gpg` and a personal dislike for how it is made. I've asked multiple questions, but I was not able to answer them based from your feedback. So: what does it take to have you not opposed to making `machinectl` working? Is the current state stripped enough from bloat, or is there some dep left which hurts a sleek base system?

shibumi commented:
@edolstra it's possible to set ncurses as pinentry for GPG. No GUI needed.

edolstra commented:
> @edolstra it's possible to set ncurses as pinentry for GPG. No GUI needed.

That's not the point: pinentry still pulls GTK+ into the closure. Or you disable at compile time, but then GUI users suffer.

> Only for those who use gpg, and then it's about +10Mbyte. 

Well bloat happens 10 MB at a time. But I don't really object to gpg in itself, the issue is everything else it drags in.

wucke13 commented:
> Well bloat happens 10 MB at a time. But I don't really object to gpg in itself, the issue is everything else it drags in.

Roughly estimated +22MB in total compared to current nixos-stable `systemd-243` (I just summed up the size from the diff in https://github.com/NixOS/nixpkgs/pull/70352#issuecomment-541010887 with the regarding sign)?

flokli commented:
@edolstra work on different pinentry flavours is happening in https://github.com/NixOS/nixpkgs/pull/49270.

wucke13 commented:
Reiterating:

+ What does it take to have you not opposed to making machinectl working? 
+ Is the current state (+22MB) stripped enough from bloat, or is there some dep left which hurts a sleek base system?
+ If these +22MB are still too much, what is your strategy of choice to solve both `machinectl` issue for the users who want it to be solved and comply with your target of a not additionally bloated base system?

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -1,5 +1,5 @@
-{ stdenv, lib, fetchFromGitHub, fetchpatch, pkgconfig, intltool, gperf, libcap, kmod
-, xz, pam, acl, libuuid, m4, utillinux, libffi
+{ stdenv, lib, fetchFromGitHub, fetchpatch, pkgconfig, intltool, gperf, libcap
+, curl, dbus, kmod, gnupg, gnutar, xz, pam, acl, libuuid, m4, utillinux, libffi
```
`dbus` doesn't seem to be referenced anywhere in the derivation.

wucke13 replied:
Will adjust soon, thank you for the notice!

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16519,6 +16519,19 @@ in
 
   systemd = callPackage ../os-specific/linux/systemd {
     utillinux = utillinuxMinimal; # break the cyclic dependency
+    dbus = dbus.override { systemd = null; }; # break the cyclic dependency
```
We don't seem to be using dbus inside systemd directly.

flokli commented:
@wucke13 with the pinentry PR merged (to staging), gpg doesn't pull in GTK or qt on the minimal installer image anymore.

You could rebase on top of that, and see how this affects closure sizes of systemd (and the minimal installation iso image).

flokli commented:
@wucke13 You're right, it didn't pull in pinentry. Sorry for the confusion.

I looked into the feature and your PR once more.

The version of gnupg we would build for systemd has the following runtime closure (I strike-though'ed all lines that were already part of systemd's runtime closure before this PR)

 - ~29M	/nix/store/hnp72ygssavd526rp7qk3yjylyc2klcw-glibc-2.27~
 - ~748K	/nix/store/1p4gc2l25ndx6pp6bpjmz0bjckwd73is-libgpg-error-1.36~
   - ~already currently pulled in via systemd's resolved~
 - ~1.3M	/nix/store/0pp7jy1kn9gpdbhnl5i205qg7hz16i15-libgcrypt-1.8.5~
 - ~140K	/nix/store/3zmbbkvxkgsp6z0qhnpdaaz7kkksq3m7-zlib-1.2.11~
 - 1.2M	/nix/store/4i67jlg1kpz3i71h39yyng5q2hyblzca-sqlite-3.28.0
   - pulled in via `${gnupg}/bin/gpg`
 - ~1.4M	/nix/store/v2x1ks46fq3yn8c94qfv38wrvmr5xn5r-bash-4.4-p23~
 - 60K	/nix/store/6jpa1h86py30i085awdhbjb6gni56rdd-npth-1.6
   - pulled in via `${gnupg}/{bin/{dirmngr,gpg-agent},libexec/scdaemon}`
 - 112K	/nix/store/hqfvc8y4bspv49fykdd6lirf6wwb6j9r-libassuan-2.5.3
   - pulled in by many tools in `${gnupg}/{bin,libexec}`
 - 268K	/nix/store/ihavkf3gwc5rmmi9dc6chypw5p363swg-libksba-1.3.5
   - pulled in via `${gnupg}/{bin/{dirmngr,gpgsm,kbxutil},libexec/scdaemon}`
 - 11M	/nix/store/kkqhf5593r5vdnlzaz9ijasrdsbh0lxm-gnupg-2.2.17

systemd basically only executes `gpg/gpg2` in `--verify` mode, in a temporary gpghome, with `--keyring` pointing to `{USER,VENDOR}_KEYRING_PATH`.

We don't need any of the other binaries in `${gnupg}/bin` (given `${gnupg}/bin/gpg` (hopefully) doesn't shell out to them by itself). 
This probably includes `dirmngr` `gpg-agent` or the smartcard functionality, so we might be able to get rid of `libksba` and `npth`.

We also don't need any of the sqlite features so we should build it with `--disable-sqlite`, saving another 1.2MiB of dependencies.

On top of that, we probably don't need to ship manpages and locales in `$gnupg/share`, shaving off another 5.1MiB, so in the end, we might end up only with `bin/gpg`, `bin/gpg2` at about 1MiB, which IMHO, is an "acceptable bloat", given we can then verify these signatures.

shibumi commented:
@flokli I can confirm. You don't need smartcard features or even a gpg-agent for the gpg support in systemd. The only purpose of this gpg keyring for systemd is too verifying gpg signed tarballs and raw gpt images. This should happen without user interaction so no smartcard support needed.

wucke13 commented:
So then, shall I add a `gnuppg-minimal` package commit to this PR with all the thinks you mentioned stripped away? @flokli

wucke13 committed:
gnugp: adding option for minimal build

Basically a flavour of gnugpg, which solely containts `bin/gnupg`.

flokli commented on pkgs/tools/security/gnupg/minimal.nix:
```diff
@@ -0,0 +1,38 @@
+    nativeBuildInputs = p.nativeBuildInputs;
+    buildInputs = p.buildInputs;
+
+    configureFlags = p.configureFlags ++ [ "--disable-sqlite" ];
```
please add this to the main gnupg derivation, like it's done with pcsclite.

wucke13 replied:
Will do

flokli commented on pkgs/tools/security/gnupg/minimal.nix:
```diff
@@ -0,0 +1,38 @@
+    bzip2 = null;
+  };
+in
+  stdenv.mkDerivation rec {
```
can't you use `overrideAttrs` together with `override` here?

Or merge the whole `postInstall` cleanup into the main gnupg derivation, behind a `enableMinimal` flag defaulting to false?

wucke13 replied:
But that way the build itself becomes fat again, since one at least has to fetch all that stuff for building a normal `gnupg`?

flokli replied:
I meant, moving the `postInstall` cleanup stuff into the gnupg derivation itself, but still calling

`gnupg.override { enableMinimal = true; pcsclite = null; sqlite = null; … }`

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3504,6 +3504,7 @@ in
   gnupg22 = callPackage ../tools/security/gnupg/22.nix {
     pinentry = if stdenv.isDarwin then pinentry_mac else pinentry;
   };
+  gnupg-minimal = callPackage ../tools/security/gnupg/minimal.nix { };
```
I don't think we should make this easily acessible in the top-level package set, better just pass the `gnupg.override {…}` into systemd'd `callPackage`, or do the override inside the systemd derivation itself.

wucke13 replied:
Which one do you prefer?

flokli replied:
personally, I'd go with defining an overridden gnupg in a let-expression inside the systemd derivation to not clutter `all-packages.nix` further…

However, people might want to pass another gnupg to build systemd, so it might make sense to do the override in `all-packages.nix`, and pass the overridden `gnupg =` down to `callPackage`.

flokli reviewed (changes requested):
Cool, gnupg now is only 1.1MiB!

Can we add some basic test of this feature into `nixos/tests/systemd.nix`?

Probably something like in upstream's test suite: `test/TEST-25-IMPORT/testsuite.sh`:
```
dd if=/dev/urandom of=/var/tmp/testimage.raw bs=$((1024*1024+7)) count=5

# Test import
machinectl import-raw /var/tmp/testimage.raw
machinectl image-status testimage
test -f /var/lib/machines/testimage.raw
cmp /var/tmp/testimage.raw /var/lib/machines/testimage.raw
```

… but with verifying a signature too.

flokli commented:
We don't have network access in the container, so just running above script during the test won't work.

We could use `fetchurl` and a known-hash to fetch images upfront, but it doesn't look like the image on that path is stable, but regularily updated, right?
Using `mkosi` would also be hard - because the bootstrapping tools it calls need network access. I guess we should just fill a bunch of random numbers like the systemd testsuite does, and sign it with a dummy key which we also configure put at systemd's `import-pubring.gpg.` We can't start this bunch of random numbers, but we can at least check the import and verification phase worked. 

In addition to the `image.tar.xz`, nspawn downloads `SHA256SUMS` and `SHA256SUMS.gpg`. I'm not sure if that's also the case if we import locally.
If it isn't, we could still cheat by enabling an nginx, creating the "image" there, and point `machinectl pull-tar` to localhost.

It might be a good idea to put these tests in a new `nixos/tests/systemd-nspawn.nix` - I think @arianvp also also has some nspawn-specific tests in the pipeline too.

wucke13 commented:
> We could use fetchurl and a known-hash to fetch images upfront, but it doesn't look like the image on that path is stable, but regularily updated, right?

Not so regularly, but there is no way to pinpoint. But the image is only about 25 MByte in size. Maybe having all of it on web archive will work?

> We can't start this bunch of random numbers, but we can at least check the import and verification phase worked.

That sounds really clever. I don't know how to do it though :laughing: 

This is the tail of output by current state of the PR:
```
SHA256 checksum of https://nspawn.org/storage/ubuntu/bionic/tar/image.tar.xz is valid.
gpg: Signature made Thu Aug  1 16:39:51 2019 UTC
gpg:                using RSA key 575DE88794A45D84456D8897A232A512E7D0BA83
gpg: Good signature from "Christian Rebischke (nspawn.org masterkey) <nspawn@nspawn.org>" [unknown]
gpg: WARNING: Using untrusted key!
Signature verification succeeded.
Created new local image 'ubuntu'.
Operation completed successfully.
Exiting.
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to start 'systemd-nspawn@ubuntu.service'.
Multiple identities can be used for authentication:
 1.  System administrator (root)
 2.  user
Choose identity to authenticate as (1-2): 2
Password:
==== AUTHENTICATION COMPLETE ====
==== AUTHENTICATING FOR org.freedesktop.machine1.shell ====
Authentication is required to acquire a shell in a local container.
Multiple identities can be used for authentication:
 1.  System administrator (root)
 2.  user
Choose identity to authenticate as (1-2): 2
Password:
==== AUTHENTICATION COMPLETE ====
Connected to machine ubuntu. Press ^] three times within 1s to exit session.
root@bionic:~#
```

flokli commented on pkgs/tools/security/gnupg/22.nix:
```diff
@@ -50,7 +50,14 @@ stdenv.mkDerivation rec {
   ] ++ optional guiSupport "--with-pinentry-pgm=${pinentry}/${pinentryBinaryPath}";
 
-  postInstall = ''
+  postInstall = if enableMinimal
```
please use `stdenv.lib.optionalString` here.

We also need to keep the symlink to `gpg2` around? We point to it inside systemd.

wucke13 replied:
Can you give me an example for that? I thought that it would append string b to string a if condition is true. I don't want that, as there is no point in copying files which shall be delete then again.

Regarding the `gpg2` binary: I don't think that's necessary
https://github.com/systemd/systemd/blob/2d901d33a90ef9d3fe01ac66c4894c9e6bf48ce0/src/import/pull-common.c#L493 calls `gpg2` first, then `gpg`. Because of that, having `gpg` only is fine, though having `gpg2` only probably fails.

flokli replied:
example for `optionalString` usage: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/system/netdata/default.nix#L43

wucke13 replied:
Ah yes. So that is not what we need there?!

flokli replied:
In the end we do something similar, but `optionalString` is more commonly used and readable IMHO.

wucke13 replied:
I'm afraid you didn't get my point (there is no point in copying files which shall be delete then again => it is  xor for the two `postInstall` strings), or I didn't understand `optionalString`. Please show me how you would change the code, then we can know for sure what we are talking about.

flokli replied:
You're right - `optionalString` probably doesn't make a lot of sense here. My bad.

flokli commented:
Can you rebase this on top of staging, change the destination branch to `staging`, and amend the 2nd commit message? Apart from that, LGTM.

wucke13 commented:
What shall be ammended with which commit message? Do you want me to squash those two commits?

flokli commented:
Not squashing is fine, the first is good on its own too.

"systemd: etc" is what's not really saying.

It could be something like:

```
systemd: add importd

This will fix the `machinectl import-.*` commands which currently fail with ...
```

wucke13 commented:
Ah I see! Done.

wucke13 commented:
@GrahamcOfBorg build systemd

flokli commented:
Can you remove the `gnupg-minimal = callPackage ../tools/security/gnupg/minimal.nix { };` addition / removal in between the two commits?

flokli committed:
nixosTests.systemd-nspawn: add test

This adds a test downloading an nspawn container via http, and ensures
sha256sum verification and gpg signature verification work.

flokli commented:
I added a systemd-nspawn test which downloads an nspawn container via http from a local server, and ensures sha256sum verification and gpg signature verification work. PTAL.

flokli commented:
@GrahamcOfBorg test systemd-nspawn

flokli commented:
Ran tests manually on both aarch64-linux and x86_64-linux.

flokli commented:
Added this test to the list in #72828 (it still needs to trickle from staging to master until it can be fixed there).

kirelagin commented:
Hey, sorry for a stupid question, but how does this work? If I am reading `meson.build` correctly, `want_importd` requires `zlib`, but it is not in `buildInputs`.

wucke13 commented:
> Hey, sorry for a stupid question, but how does this work? If I am reading `meson.build` correctly, `want_importd` requires `zlib`, but it is not in `buildInputs`.

I guess things which are not explicitly deactivated are enabled on demand!

kirelagin commented:
The thing is in that PR you set `-Dimportd=true`, which activates [this `if`](https://github.com/systemd/systemd/blob/v243/meson.build#L1240) and then, if `zlib` is not found, [this `if`][https://github.com/systemd/systemd/blob/v243/meson.build#L1245) should trigger and configure should fail with an error. And `HAVE_ZLIB` [is only ever set to true](https://github.com/systemd/systemd/blob/v243/meson.build#L1091) if `zlib` was actually found.

kirelagin commented:
Ok, it ends up being in the dependencies closure _somehow_, probably as a propagated input of some other package, huh.

adrianparvino commented:
Hi, this caused a regression on our machines.

58fb23f72ad
```
exchange.> reloading the following units: dbus.service, firewall.service
```

6317f5b81ce
```
exchange.> stopping the following units: dbus.service, dbus.socket, nscd.service
exchange.> Failed to create bus connection: Connection refused
```

flokli commented:
I think this is the same issue as https://github.com/NixOS/nixpkgs/issues/74626 and not immediately caused by this PR.
